### ISSUE TITLE
Reproducibility attestation

### ISSUE THREAD

**Issue by basantagoswami** [zapstore/zapstore#23]
A user (dev) should be able to declare that a particular version of any app is reproducible _(or not)_ from a particular commit or release tag. Show warnings when a dev in your WoT says they have failed to generate a reproduced build.

**Comment by franzaps** [comment]
Yes, planned

**Comment by franzaps** [comment]
Reproducible build DVMs?

**Comment by Giszmo** [comment]
Reproducible builds are super ultra niche and non-devs don't get it. Even devs often have a hard time understanding the implications. Therefore it's an up-hill battle to get people to care enough to get people to produce these builds.

In the context of WalletScrutiny, we not only attest to reproducibility but also attest to failed reproducibility which I find almost more important than the former. If one can reproduce, great. If all can reproduce, perfect. If one fails, it might be just incompetence but if several agree to not be able to RB, it might hint at some serious problem if not backdoor in a build tool for example. Therefore, many RB make sense and it makes sense for all of them to report back and it's usually not trivial to RB artifacts. Once setup, automation is possible but who will pay for it? Granted, I would but I would not pay for clones of the same test getting to the same result or to some anonymous account that might be lying in favor of one or several projects.

So yes, maybe some people might offer RB as a service for artifacts uploaded to bloom servers but I personally don't see the DVM potential there. I hope I'm wrong and we get tons of RBs.

**Comment by alltheseas** [comment]
Why not keep it simple: hard core RB maxis can sign a self attestation that they have achieved a RB for a particular app, or not.

An app can therefore have npubs who:
-follow the app's npub
-recommend the app
-verify the app's SHA256 hash (or not)
-verify RB (or not)

With permissionless zaps maybe there is a way to incentivize the latter.

**Comment by Giszmo** [comment]
@alltheseas I don't see your argument or in what it differs what was said above.

Another question is: What constitutes "an app"? Who decides which one is "the app's npub"? 

At WalletScrutiny, an app is defined as for example the appId on Google Play. The provider makes many releases and maybe multiple releases per version, which all constitutes artifacts of the app which would need to be RB tested. For Desktop apps it gets really complicated. Bitcoin Core for example can be downloaded from dozens of different pages or repositories, 20 of which are even referenced on bitcoincore.org and then there is the bitcoin/bitcoin github repo. Is snapcraft/bitcoincore one app? Or are debian/bitcoincore-amd46 and debian/bitcoincore-32 different apps? Or are these all artifacts of the same app? If so, how does recommending an app work if for example the app on gentoo blacklists some addresses? Should the user be bothered with such nuance? Can he be not bothered with it, without a recommendation for an app driving people to install the one artifact that is outright evil?

**Comment by alltheseas** [comment]
My point was not everything needs to be a DVM. 

> What constitutes "an app"? Who decides which one is "the app's npub"?

> Is snapcraft/bitcoincore one app? Or are debian/bitcoincore-amd46 and debian/bitcoincore-32 different apps? 

Good questions. FOSS x WOT will come in handy here.

**Comment by franzaps** [comment]
Re: attesting to failed reproducibility, agree is more important. 

The market will determine how many RB DVMs we get. Would you like to become one? You have a reputation you can leverage. Are your RBs today published as some nostr event?

> What constitutes "an app"? Who decides which one is "the app's npub"?

An app is a [parameterizable replaceable event of kind 32267](https://github.com/zapstore/zapstore/wiki/Sample-app-events). Its `d` tag is the app identifier (easy for APKs).

Anyone can publish an app – even if it's not theirs. Web of trust, curation, badges, DVMs will become essential here. `debian/bitcoincore-amd64` and `debian/bitcoincore-32` could be published as one app with two different architectures (`arch` tag in kind 1063) or as completely separate apps by different pubkeys. Our (non-trivial) job is to surface and give tools to users to help them determine which app is the right one for them.

An app's npub is the npub signing the 32267 event.

**Comment by Giszmo** [comment]
> Would you like to become one?

I don't know. I see a tragedy of the commons here and I'm not sure a DVM is the right thing here. So lets say, somebody pays for an RB test of wallet X. I can't deliver instantly and I would rather publish than to DM the result, so would the DVM show the user: sats paid for this artifact's RB test: 5000 since yesterday? I could imagine to sell minutes trying to RB test stuff and sure, with 5h paid I'd have to dig deeper than with 5min paid. I'm also not familiar with other DVMs or how they integrate with clients.

> Are your RBs today published as some nostr event?

I want to do that. They eventually will be as we tackle desktop with its dozens of artifacts per version.

**Comment by franzaps** [comment]
I get your point. You make build results public anyway, and requesting an app to be built on the fly likely won't work.

I definitely see developers linking their 1063s to an attestation with matching hash signed by Wallet Scrutiny's npub. That can really push their apps up the trust ranking.

**Comment by franzaps** [comment]
F-Droid reproducible builds
 - https://verification.f-droid.org/

**Comment by Giszmo** [comment]
F-Droid attestations are nice to have but they are not cryptographically signed and it's not clear **who** attests to RB. I think attestations on nostr will add something special there.

**Comment by franzaps** [comment]
@Giszmo have you found a NIP for these attestations?

**Comment by Giszmo** [comment]
@franzaps no. I'm a bit in analysis paralysis. If you have something, it would be very much appreciated to move forward on this. It's urgently needed by WalletScrutiny.

**Comment by franzaps** [comment]
@Giszmo can you share here some basic attestation data, in some JSON format? We can then agree on something and call it a NIP

**Comment by Giszmo** [comment]
https://github.com/Giszmo/nips/blob/2413696aed7f54ba110d1652acec6c8bada45399/attestations.md is a first draft.

**Comment by franzaps** [comment]
@Giszmo Looks good, quick feedback:

Why not make this specific to software attestations rather than generalized for tomatoes and whatnot?

>  t: MUST be set to "attestation"

Why is this, if the kind already defines it as an attestation?

**Comment by Giszmo** [comment]
I agree. Let's not over-complicate it. Let others figure out how to attest to tomato farms.

For software I made some improvements. Your opinion would be appreciated.

**Comment by auggie-lahey** [comment]
since we have working CICD dvms, i'd like to open this conversation back up start offering this service. 

`{
  "kind": 31000,
  "content": "I, John Doe, attest that the amethyst apk with sha256 hash 1345234231 can be reproduced from the source code revision 345213451",
  "tags": [
    ["d", "amethyst artifact 1345234231"],
    ["t", "attestation"],
    ["c", "reproducible"],
    ["git", "345213451"],
    ["version", "25.2-beta"],
    ["evidence", "a:<my pub key>:30023:BC_1345234231"],
    ["evidence", "link to DVM run"],
    ["evidence", "link to blossom upload of artifact from Dev"],
    ["evidence", "link to blossom upload of artifact from RB DVM"]
  ]
}
`

would this be what you're looking for when using amethyst as an example? would you want that published to the zap.store relay?

**Comment by franzaps** [comment]
Nice. I see a lot of superfluous information there, or am I wrong? The content, especially. Kind, git commit (and maybe version) and hash should be enough. Is evidence needed? I am trusting your reputation anyway.

@Giszmo you had come up with a spec a few months ago, did you move forward with that?

**Comment by franzaps** [comment]
Also: it's Zapstore (not zap.store)

**Comment by Giszmo** [comment]
I think we have to try it out but I want to have room to elaborate on details so a special way of linking a long-form article makes sense. It could be optional.
One of my headaches is how to handle multiple artifacts. In the case of Android apps we are dealing with Android App Bundles where the Android phone downloads multiple artifacts that would not have one natural hash to them so attestations would be to base.apk and en.apk and hdpi.apk with 3 hashes or it would be 3 attestations.

@ktecho any comments on this?

**Comment by auggie-lahey** [comment]
{
  "kind": 31000,
  "content": "I, John Doe, attest that the amethyst apk with sha256 hash 1345234231 can be reproduced from the source code revision 345213451",
# tbh, the actuall attestation content for me is irrelevant because its just a summary of the below tags
  "tags": [
    [
      "d",
      "amethyst artifact 1345234231" # app name and commit, could be separate
    ],
    [
      "t", 
      "attestation"
    ],
    [
      "status", # changed from c to status
      "reproducible"
    ],
    [
      "git",
      "345213451" # reference to commit
    ],
    [
      "version", # version of the software
      "25.2-beta"
    ],
    [
      "evidence",
      "a:<my pub key>:30023:BC_1345234231"
    ],
    [
      "evidence",
      "link to DVM run" # our intention is to run these using our new dvms that run cicd pipelines from any github workflow
    ],
    [
      "evidence",
      "link to blossom upload of artifact from Dev" # this basically provides the hash of the artifact from the dev
    ],
    [
      "evidence",
      "link to blossom upload of artifact from RB DVM" # this provides the hash of the artifact from the RB pipeline which should match that which the dev gives. 
    ]
  ]
}****

**Comment by auggie-lahey** [comment]
if you think the evidence piece is overkill, i can basically put all of that  in the content. But ideally our plan is the following workflow

- dev pushes release
- release triggers dvm
-store release on blossom A
- dvm run workflow and publishes artifact to blossom B
- hashes should be matching, if not irreproducible

if we publish the evidence, then devs would be able to walk through the pipeline to figure out what the issue is if a build randomly turns up irreproducible

**Comment by franzaps** [comment]
This is great, thank you.

We have two main pieces here: the event describing the artifact, and the event describing the attestation.

### Artifact event

@ktecho is using kind 106300, @auggie-lahey is using none(?) and Zapstore is using kind 1063 (NIP-94)

I chose [NIP-94](https://github.com/nostr-protocol/nips/blob/master/94.md) because it's an established, merged NIP and it's widely used including Zapstore, with thousands of such events.

(To be fair, I extended it slightly to include the app ID, version, APK signature hash, etc – I will include an example below)

Could we all agree to use NIP-94 or is there any objection?

A remaining question though is: How do we "normalize" the source of the artifact? We can use app IDs only but adding a source can prevent confusion (as anyone could use any app ID for their app without enforcement obviously). One could think of URLs, but then artifacts coming from Google Play Store don't really have a URL. Maybe define a few standardized sources: `googlePlay`, `fdroid`, etc.  Open to other ideas on this one.

### Attestation event

@ktecho is using kind 303010, @auggie-lahey is using kind 31000

@auggie-lahey the hash in the `d` tag mixed with other information doesn't seem very helpful. Do you mind doing like @ktecho and using the `x` tag? What's nice, is `x` is queryable in relays and is easily linkeable to the artifact event (NIP-94?) which also has an `x` tag.

@ktecho any reason why you don't include the git commit in the tags like @auggie-lahey does?

For `status` there seems to be an agreement? So using `reproducible`, and I presume `irreproducible` could be the other possible value?

In terms of evidence tags and other info pertaining to the build I don't have an opinion, so whatever you guys think is best.

In the Zapstore UI, so you get an idea, I would want to show a badge when the APK has been attested as reproducible (or otherwise). I would query your relays to find attestation events for `x` hash, and tap to display more information about the attestation.

## AABs

@Giszmo raises an interesting one as well about AABs – I definitely agree base.apk and other APKs should have separate artifact + attestation events. We could add an `artifactName` property or similar to the artifact event (NIP-94?) to describe which specific APK, e.g. `base.apk`, the artifact event is making reference to, as all these APKs would share the exact same app ID, version, developer, etc.

## Example of (extended) NIP-94 event

This one comes from the Zapstore relay and is signed by Zeus.

```json
{
    "id": "5f48f81f1bf0aad89aa6465e46b3c05eef9abd93ac32635d870b449e13ce3925",
    "pubkey": "34d2f5274f1958fcd2cb2463dabeaddf8a21f84ace4241da888023bf05cc8095",
    "sig": "6841813c0463ccab21285c336cbed3eaa584076f17381e2d712d8fc974c508629b2bf53896475b9c82f13ebcb75c8b7e97a1a72d688c8e2da7f8824a6657bd94",
    "kind": 1063,
    "created_at": 1739057947,
    "content": "app.zeusln.zeus@0.9.6",
    "tags": [
        [
            "version_code",
            "99002"
        ],
        [
            "min_sdk_version",
            "28"
        ],
        [
            "target_sdk_version",
            "34"
        ],
        [
            "apk_signature_hash",
            "2af8e20ac9445767cbd44ed84dbbfc33c6c98248897c4f843c42c2765c4ad3ba"
        ],
        [
            "url",
            "https://cdn.zapstore.dev/3b297604b7e337be56fa1239a5b9f4f33480f71718703fa8abc53f51a344dc2a"
        ],
        [
            "m",
            "application/vnd.android.package-archive"
        ],
        [
            "x",
            "3b297604b7e337be56fa1239a5b9f4f33480f71718703fa8abc53f51a344dc2a"
        ],
        [
            "size",
            "61998491"
        ],
        [
            "version",
            "0.9.6"
        ],
        [
            "f",
            "android-x86"
        ]
    ]
}
```


Hope all this made sense. Would be awesome to agree on a standard and spec it out.

**Comment by Giszmo** [comment]
I think @ktecho took an example with a temporary kind number, so the artifact event would use the nip94 one as hinted at with the same initial digits.

The source of the artifact could be one or several URLs. There should always also be a content addressable source like a blossom server, too. Even if the file came from Google, pushing a copy to a blossom server should help to not lose rogue artifacts.

I think the attestation event in our PR is also in the kind 30k range. 

A git commit should be optional.

Status should be extendable but for a start:

* `reproducible`
* `irreproducible`
* `ftbfs`, failed to build from source
* `notag`, The git revision to compile is not clear
* `nosource`, The source for this version was not found or the repository was taken down
* `warning`, This could be used if the binary is reproducible but some other red flag has come up

**Comment by ktecho** [comment]
@franzaps yes, @Giszmo is right. The correct kinds are the ones stated in the documentation (`1063` and `30301`). The other ones are our `beta` kinds to not polute the official kinds with testing info.

We aren't using the official NIP-94 because it mandates certain fields that we don't have. Take for example the URL (we may or may not have one). Also, `x` and `ox` have duplicated information for our use case. We're sending both right now, but I think we may leave just one when I have the time to think about it better :)

**Comment by franzaps** [comment]
@ktecho @Giszmo 

Thanks for this. You mention a PR and documentation, can you share links to those?

In NIP-94, just writing `x` and an empty `url` should be fine. As for git commits, I get what you say, but when you do attest from a git commit what do you use?

So how does an attestation to the 1063 event from Zeus I shared above look today? Can we start moving forward with this so I can start displaying these attestations on Zapstore?

@auggie-lahey do you think we could use the same standard in the CI/CD pipelines project?

**Comment by Giszmo** [comment]
@franzaps here are [documentation](https://gitlab.com/walletscrutiny/walletScrutinyCom/-/blob/master/doc/attestations.md) and [merge request](https://gitlab.com/walletscrutiny/walletScrutinyCom/-/merge_requests/924).

I agree that rather than using a different kind because we are not precisely on target we should use the wider recognized kind.

WalletScrutiny works the other way around to many other "reproducible builds" projects. Where others treat reproducibility as an aspect of the code, we treat it strictly as an aspect of a binary blob and that doesn't always come with an explicit or implied source code. As I mentioned above, the status could be `nosource` or `notag`, both of which would make quoting a git commit hard. If we do have a git commit, we should provide it.

**Comment by auggie-lahey** [comment]
we can use whatever you want. i'm not super familiar with the different d or x tags in the nip repo so those are random in my comment. 

If you want me to try to get one running, just respond with an updated event from my comment to what you want and i can try to implement that soon.

**Comment by franzaps** [comment]
>  If we do have a git commit, we should provide it

Yes, exactly

@auggie-lahey can you see if the attestation event [here](https://gitlab.com/walletscrutiny/walletScrutinyCom/-/blob/master/doc/attestations.md) works for you?

@Giszmo @ktecho are you interested in having your attestations show up in Zapstore? Do you happen to verify Zeus as I gave in the example (and hopefully others too)? If so I could query your relay for that.

**Comment by Giszmo** [comment]
> [@Giszmo](https://github.com/Giszmo) [@ktecho](https://github.com/ktecho) are you interested in having your attestations show up in Zapstore?

We are using nostr precisely because we want to be interoperable and I reached out to you because I hope we get something the community can get behind, so yes, of course, we want that!

> Do you happen to verify Zeus as I gave in the example (and hopefully others too)? If so I could query your relay for that.

Yes, Zeus is one of the apps we check for RB. @ktecho will publish legacy attestations as nostr events but we are still in beta, so ... patience, please 🙏

**Comment by franzaps** [comment]
That's awesome. Keep me updated.

**Comment by Giszmo** [comment]
Some things are now out in production. You can check for example https://walletscrutiny.com/verifier/?pubkey=1f9e547c2f31942623b8ad1d07713282e8640fd8cf474e9f79f18ace8af216ed

It's all still a bit rough around the edges but we (@ktecho) are smoothing it out quickly.

**Comment by franzaps** [comment]
> Some things are now out in production. You can check for example https://walletscrutiny.com/verifier/?pubkey=1f9e547c2f31942623b8ad1d07713282e8640fd8cf474e9f79f18ace8af216ed
> 
> It's all still a bit rough around the edges but we ([@ktecho](https://github.com/ktecho)) are smoothing it out quickly.

I was expecting a nostr event for the attestation, is this somewhere on the page?

**Comment by ktecho** [comment]
> > Some things are now out in production. You can check for example https://walletscrutiny.com/verifier/?pubkey=1f9e547c2f31942623b8ad1d07713282e8640fd8cf474e9f79f18ace8af216ed
> > It's all still a bit rough around the edges but we ([@ktecho](https://github.com/ktecho)) are smoothing it out quickly.
> 
> I was expecting a nostr event for the attestation, is this somewhere on the page?

Each line of that table is an attestation. Also, if you go to one of the wallet pages, you'll also see the attestations for that wallet. For example, for Zeus ([link](https://walletscrutiny.com/android/app.zeusln.zeus/)), you'll get this:

![Image](https://github.com/user-attachments/assets/57cf3e31-c0f5-45dd-9d84-a3ba50156f8d)

**Comment by franzaps** [comment]
> > It looks awesome, congrats. But in the context of Zapstore - how do I integrate this? Do I query your relay, hit a DVM?
> 
> We're using 6 public relays, some of them in America, others in Europe. I guess you could ask them :)
> 
> https://gitlab.com/walletscrutiny/walletScrutinyCom/-/blob/master/src/nostr-constants.mjs#L3

You gonna make me work eh? 😅

So you have no spec for the nostr events? NIP drafts?

Let's say I'm on the app detail screen, I understand I should query for kind=32304 and x=hash and check for `status`.

@auggie-lahey are you on the same page here with the nostr kind?

**Comment by ktecho** [comment]
> So you have no spec for the nostr events? NIP drafts?

We do:

https://gitlab.com/walletscrutiny/walletScrutinyCom/-/blob/master/doc/verifications.md?ref_type=heads

Endorsements are not implemented yet, so the spec could change in that respect.

**Comment by franzaps** [comment]
@ktecho sorry I don't quite get the difference between verification and endorsement. Which one should I query for? Why do binaries have multiple hashes?

**Comment by Giszmo** [comment]
@franzaps please take our approach apart. It's early days and this is what we came up with. Especially the multiple hashes might feel weird but for us it makes sense. So when testing the reproducibility of a signed binary, the main hash might be the signed binary but we include the nulled signature file hash, too, as the hardware wallet shows that one etc.

**Comment by franzaps** [comment]
@Giszmo in general it looks good to me. I noticed there were updates and the URL changed to: https://gitlab.com/walletscrutiny/walletScrutinyCom/-/blob/master/docs/verifications.md

I noticed you are using the `platform` tag in 1063 events. I understand the need for that, I'm using the `f` tag for about a year now. The platforms are well-defined standardized strings to make it easier to interoperate. Do you think you could use them too? Here: https://github.com/franzaps/nips/blob/applications/82.md#platforms

I don't quite understand the need for endorsements, as the work for a verification and an endorsement is the same. But now clients need to query for two kinds instead of one, and builders need to search for verifications to know whether they should publish an endorsement or a verification?  From the perspective of Zapstore, I want to query for kind=30301 on a bunch of relays and have results from WalletScrutiny and other providers, that's it.

About the multiple hashes, I understand the problem. Need to think about it more.

**Comment by franzaps** [comment]
On another note, @Giszmo and @ktecho , Vitor Pamplona suggested to me that a specific kind for Software assets be created. As we are extending 1063s with lots of custom tags. And that it should be defined in this NIP too: https://github.com/franzaps/nips/blob/applications/82.md . Maybe we could come up with it together.

**Comment by ktecho** [comment]
> I don't quite understand the need for endorsements, as the work for a verification and an endorsement is the same.

- `Verifications` are made by people that know a bit more about how to compile an app, how to install libraries, how to create a Dockfile, or how to debug a problem while building. So they do that and tell the world what they did and the result of that process.
- `Endorsements` are just people that know a bit less, but can follow steps outlined in a previous verification to confirm that the verification is correct. It's like a `+1` or a `like` in a social network, but with a bit more of weight, because you've supposedly repeated the process and you've come to the same conclusion.

Having said that, Endorsements are still a think we're thinking about, and they're not implemented in any way yet. In fact, I'll remove them from that doc so we don't misslead other people.


> I want to query for kind=30301 on a bunch of relays and have results from WalletScrutiny and other providers, that's it.

Yes, you can already do that, and doesn't matter if we finally implement Endorsements or not, it will have the same value, as Endorsements don't have identity on themselves. They'll only give more `weight` to an already existing verification, so if you ignore them, you will still show what you want: verifications.

**Comment by franzaps** [comment]
Thank you @ktecho !  Can you let me know your thoughts on the 1063s?

**Comment by alltheseas** [comment]
When I re-read this it seems very technical to be considered at scale by many. 

I presume most people would delegate their trust to e.g. gigi, Max, Leo etc. to think about these technical, difficult things on their behalf.  #266 "catalog level trust" comes to mind.

This feature is closer to belonging in a technical/security report than a (B2)B2C app in Zapstore.

It is certainly incredibly valuable information, an upgrade to how all trust goes into central appstores today. 

How do we make use of this, whilst not scaring away folks 

cc @NielLiesmons

**Comment by Giszmo** [comment]
So @alltheseas you suggest the app could have endorsements per artifact? Or what is the catalog level trust exactly? The issue's description sounds quite like the opposite.
If you would have artifact specific endorsements it could be agnostic as to why these endorsements exist and we could use it to endorse reproducible artifacts.
